tests/zfs-tests/include/kstat.shlib

```diff
@@ -0,0 +1,507 @@
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright (c) 2025, Klara, Inc.
+#
+
+#
+# This file provides the following helpers to read kstats from tests.
+#
+#   kstat [-g] <stat>
+#   kstat_pool [-g] <pool> <stat>
+#   kstat_dataset [-N] <dataset | pool/objsetid> <stat>
+#
+# `kstat` and `kstat_pool` return the value of of the given <stat>, either
+# a global or pool-specific state.
+#
+#   $ kstat dbgmsg
+#   timestamp    message
+#   1736848201   spa_history.c:304:spa_history_log_sync(): txg 14734896 ...
+#   1736848201   spa_history.c:330:spa_history_log_sync(): ioctl ...
+#   ...
+#
+#   $ kstat_pool garden state
+#   ONLINE
+#
+# To get a single stat within a group or collection, separate the name with
+# '.' characters.
+#
+#   $ kstat dbufstats.cache_target_bytes
+#   3215780693
+#
+#   $ kstat_pool crayon iostats.arc_read_bytes
+#   253671670784
+#
+# -g is "group" mode. If the kstat is a group or collection, all stats in that
+# group are returned, one stat per line, key and value separated by a space.
+#
+#   $ kstat -g dbufstats
+#   cache_count 1792
+#   cache_size_bytes 87720376
+#   cache_size_bytes_max 305187768
+#   cache_target_bytes 97668555
+#   ...
+#
+#   $ kstat_pool -g crayon iostats
+#   trim_extents_written 0
+#   trim_bytes_written 0
+#   trim_extents_skipped 0
+#   trim_bytes_skipped 0
+#   ...
+#
+# `kstat_dataset` accesses the per-dataset group kstat. The dataset can be
+# specified by name:
+#
+#   $ kstat_dataset crayon/home/robn nunlinks
+#   2628514
+#
+# or, with the -N switch, as <pool>/<objsetID>:
+#
+#   $ kstat_dataset -N crayon/7 writes
+#   125135
+#
+
+####################
+# Public interface
+
+#
+# kstat [-g] <stat>
+#
+function kstat
+{
+	typeset -i want_group=0
+
+	OPTIND=1
+	while getopts "g" opt ; do
+		case $opt in
+			'g') want_group=1 ;;
+			*) log_fail "kstat: invalid option '$opt'" ;;
+		esac
+	done
+	shift $(expr $OPTIND - 1)
+
+	typeset stat=$1
+
+	$_kstat_os 'global' '' "$stat" $want_group
+}
+
+#
+# kstat_pool [-g] <pool> <stat>
+#
+function kstat_pool
+{
+	typeset -i want_group=0
+
+	OPTIND=1
+	while getopts "g" opt ; do
+		case $opt in
+			'g') want_group=1 ;;
+			*) log_fail "kstat_pool: invalid option '$opt'" ;;
+		esac
+	done
+	shift $(expr $OPTIND - 1)
+
+	typeset pool=$1
+	typeset stat=$2
+
+	$_kstat_os 'pool' "$pool" "$stat" $want_group
+}
+
+#
+# kstat_dataset [-N] <dataset | pool/objsetid> <stat>
+#
+function kstat_dataset
+{
+	typeset -i opt_objsetid=0
+
+	OPTIND=1
+	while getopts "N" opt ; do
+		case $opt in
+			'N') opt_objsetid=1 ;;
+			*) log_fail "kstat_dataset: invalid option '$opt'" ;;
```

Comment:
We might not use it now, but wouldn't it be better to support `-g` here also for uniformity?

Reply:
I didn't, because the `objset-0xnnn` is already a group stat, and in this model, a group can't have a group in it.

My first cut of this let you leave the stat out, and then it would give you the whole group, but I didn't like the asymmetry - it seemed weird that `zpool_dataset` had an optional stat, but the others didn't. So I decided to just keep it tight, limit it to what we actually use for now.

This whole exercise has got me thinking a lot about how the whole stats model should be redone. I suspect in the some future the dataset will be a real "kind", with multiple group stats under it. Then we can add `-g` :)

Reply:
> I didn't, because the `objset-0xnnn` is already a group stat, and in this model, a group can't have a group in it.

Ah.  Whoever first designed the dataset stats could have added another layer there to allow further nesting.  FreeBSD does not have this problem actually, allowing values and further nesting at the same level.